lib/attr_extras/attr_initialize.rb

```diff
@@ -12,22 +12,13 @@ def apply
     names = @names
     block = @block
     validate_arity = method(:validate_arity)
-    set_ivar_from_hash = method(:set_ivar_from_hash)
+    set_ivar_from_args = method(:set_ivar_from_args)
 
     klass.send(:define_method, :initialize) do |*values|
       validate_arity.call(values.length, self.class)
 
       names.zip(values).each do |name_or_names, value|
```

Comment:
Yeah, that's a rabbit hole. I'd need to zip those positionals and optional values separately. I was thinking about something along the lines of:

```ruby
names, array_names = names.partition{ |n| !n.is_a? Array }  
values, hash_values = values.partition{ |n| !n.is_a? Hash }
names.zip(values) + array_names.zip(hash_values)
```

as for the arity it's probably needed to split those name hashes:
```ruby
symbol_names, hash_names = names.partition{ |n| !n.is_a? Hash }
@names = symbol_names + hash_names.inject(:+).map{ |k, v| { k => v } }
```
or sum it somehow
```ruby
arity_with_hashes = names.map{|x| x.is_a?(Hash) ? x.count : 1 }.inject(:+)
```

Thoughts?